File: 053 Maximum Subarray/maximum_subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        global_max = nums[0]
        max_end_here = nums[0]
        for i in range(1, len(nums)):
            max_end_here = nums[i] if nums[i] > max_end_here + nums[i] else max_end_here + nums[i]
            if max_end_here > global_max:
                global_max = max_end_here
        return global_max

File: 53 Maximum Subarray/MaximumSubarray.java
class MaximumSubarray {
    public int maxSubArray(int[] nums) {
        if (nums == null || nums.length == 0) {
            return -1;
        }
        int maxEndHere = nums[0];
        int globalMax = nums[0];
        for (int i = 1; i < nums.length; i++) {
            maxEndHere = nums[i] > nums[i] + maxEndHere? nums[i]: nums[i] + maxEndHere;
            if (maxEndHere > globalMax) {
                globalMax = maxEndHere;
            }
        }
        return globalMax;
    }
}